### DISCUSSION TITLE
Some bounding boxes don't move on: center (pivot:above)

### DISCUSSION THREAD

**Discussion by bigrig2212** [donmccurdy/glTF-Transform#1018] [Q&A] [answered]
Hi - i've been doing recentering to assets in batch and discovered that some assets' bounding boxes get left behind. 

<img width="675" alt="Screenshot 2023-07-08 at 3 07 30 PM" src="https://github.com/donmccurdy/glTF-Transform/assets/19869387/629b7bf3-6790-479a-8023-b93d97c9b819">

Suggestions? 

(and potentially related - after both centering and optimizing, some of these meshes are missing in the mesh hierarchy.... like i can't find them anymore in gltf viewer - or they get renamed to "nodeN" or they simply dont show up as addressable anywhere ... i figure this is by design in the optimization process? Or maybe related to this same behavior with the centering bounding boxes?)  the impact is that meshes disappear on camera orbit. im assuming b/c the bounding box is getting frustum culled - even though the mesh itself is still in frustum.

**Comment by bigrig2212** [comment]
Here is the original, pre-transform:
<img width="627" alt="Screenshot 2023-07-08 at 4 13 08 PM" src="https://github.com/donmccurdy/glTF-Transform/assets/19869387/bc33279b-abdc-43d5-99bb-be8a3767d21b">

**Reply by donmccurdy** [reply]
@bigrig2212 to help here, I think I'd need information about:

1. What glTF files are you testing with?
2. What operations are you doing on those files?

> some of these meshes are missing in the mesh hierarchy..

If one of the optimizations you're doing happens to be `gltf-transform optimize ...` then this is expected. It attempts to flatten the scene graph and join meshes where possible, so it's normal that the output has fewer meshes and nodes than the input. You can disable that with options `--no-flatten` and `--no-join`.

**Reply by bigrig2212** [reply]
Sure thing! 
1. Here's the source file and the compressed version attached. 
[west.models.zip](https://github.com/donmccurdy/glTF-Transform/files/11996016/west.models.zip)

2. the operations i ran are these: 
gltf-transform center "$file" "output/$filename" --pivot 'below'
gltf-transform optimize "$file" "output/$filename" --join false --flatten false --simplify 0.0

With join/flatten to false and simplify at 0, i thought i'd prevent joined meshes (i need them separate for my app, which enables mesh picking from the scene). 

appreciate you taking a look!!

**Reply by donmccurdy** [reply]
Ahh ok. It looks like instancing is what's causing the remaining change — if you add the `--verbose` flag you'll see this output:

> info: instance: Created 32 batches, with 573 total instances.

So that's saving you 500+ draw calls, using the glTF extension `EXT_mesh_gpu_instancing`. This may or may not be important depending on your goals, but it does mean the output has fewer nodes than the original, even without explicitly flattening the scene graph. If you'd prefer to avoid that, try:

```bash
gltf-transform optimize "$file" "output/$filename" --no-join --no-flatten --no-simplify --no-instance
```

The file size remains very small, but the draw call count will be higher.

**Reply by bigrig2212** [reply]
Saving draw calls is amazing (as is file size reduction).

Is there a combination of options to maintain instancing along with mesh hierarchy and bounding boxes around each instance (eg: to enable picking of each discrete mesh and/or instance)

**Reply by donmccurdy** [reply, marked as answer]
If GPU instancing is baked into the file, then each batch of meshes (e.g. all identical trains are one batch) is represented as a single node in the scene. Bounding boxes and picking are a viewer/engine concepts — not inherent in the file. You can certainly _implement_ bounding boxes and picking for files with GPU instancing, but I don't know how/if the Babylon.js sandbox viewer handles this.

It's also possible to skip baking instancing into the file, but still allocate GPU instances at runtime instead. If you're using React Three Fiber (https://github.com/pmndrs/react-three-fiber) then I think it has some tools to automate that, or you can write the code to create instances. Doesn't have to be done ahead of time. But this still requires that you implement picking in a way that works with the instancing.

**Reply by bigrig2212** [reply]
ah - this all makes perfect sense now. thanks for the tut! learned a lot from this.

**Reply by bigrig2212** [reply]
btw, when i try your syntax above i get errors. need to write command as: 
gltf-transform optimize "$file" "output/$filename" --join false --flatten false --simplify 0.0 --instance 0.0

although its still spending time in the instance step - hm - actually looks like its still instancing and creating nodes..

help says:
```
--instance <min>                     Create GPU instances from shared mesh references       
                                         number, default: 5
```

Seems like setting it to 100 results in no nodes being created. 
What does "Min" mean in this case? 

Oh i see:
```
info: instance: No meshes with ≥100 parent nodes were found.
debug: instance: Complete.
✔ instance             3ms
```
it's depth level. Got it.

**Reply by bigrig2212** [reply]
Wondering: how do i turn it off all together?

**Reply by donmccurdy** [reply]
It sounds like you might have an older version of the CLI installed – does re-running `npm install --global @gltf-transform/cli` resolve that?

**Reply by bigrig2212** [reply]
Ah! I'd updated all of the package.json libraries ... didnt think about the cli itself. Yep, that fixed. Thx!
